core/src/main/java/org/apache/iceberg/io/DeleteSchemaUtil.java

```diff
@@ -20,13 +20,38 @@
 package org.apache.iceberg.io;
 
 import org.apache.iceberg.MetadataColumns;
+import org.apache.iceberg.Partitioning;
 import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.types.TypeUtil;
 import org.apache.iceberg.types.Types;
 
 public class DeleteSchemaUtil {
   private DeleteSchemaUtil() {
   }
 
+  public static Schema metadataTableSchema(Table table) {
+    return metadataTableSchema(table, Partitioning.partitionType(table));
+  }
+
+  public static Schema metadataTableSchema(Table table, Types.StructType partitionType) {
+    Schema result = new Schema(
+        MetadataColumns.DELETE_FILE_PATH,
+        MetadataColumns.DELETE_FILE_POS,
+        Types.NestedField.optional(
+            MetadataColumns.DELETE_FILE_ROW_FIELD_ID, "row", table.schema().asStruct(),
+            MetadataColumns.DELETE_FILE_ROW_DOC));
+
+    if (partitionType.fields().size() > 0) {
```

Comment:
I feel like using `TypeUtil.selectNot` will make it easier to read:

```
Types.StructType partitionType = Partitioning.partitionType(table);
Schema schema = new Schema(...);

if (partitionType.fields().size() < 1) {
  // avoid returning an empty struct, which is not always supported. instead, drop the partition field
  return TypeUtil.selectNot(schema, Sets.newHashSet(PARTITION_ID));
} else {
  return schema;
}
```

Reply:
This was emulated from other metadata table like BaseFilesTable, but good point.